src/state.rs

```diff
@@ -23,6 +24,8 @@ pub enum ReplicaState {
 pub struct LeaderState {
     next_index: HashMap<ServerId, LogIndex>,
     match_index: HashMap<ServerId, LogIndex>,
+    /// Stores in-flight client proposals.
+    pub proposals: VecDeque<(ClientId, LogIndex)>,
```

Comment:
Awesome, this is what I was thinking too. A `VecDeque` is a good structure for this.